CesiumGltfContent/src/GltfUtilities.cpp

```diff
@@ -901,4 +1228,265 @@ void GltfUtilities::compactBuffer(
+    const CesiumGeometry::Ray& ray,
+    const CesiumGltf::Model& model,
+    const CesiumGltf::MeshPrimitive& primitive,
+    const Accessor& positionAccessor,
+    const glm::dmat4x4& primitiveToWorld,
+    bool cullBackFaces,
+    std::vector<std::string>& warnings) {
+  glm::dmat4x4 worldToPrimitive = glm::inverse(primitiveToWorld);
+  CesiumGeometry::Ray transformedRay = ray.transform(worldToPrimitive);
+
+  // Ignore primitive if ray doesn't intersect bounding box
+  const std::vector<double>& min = positionAccessor.min;
+  const std::vector<double>& max = positionAccessor.max;
+
+  double t;
+  if (!CesiumGeometry::IntersectionTests::rayAABBParametric(
+          transformedRay,
+          CesiumGeometry::AxisAlignedBox(
+              min[0],
+              min[1],
+              min[2],
+              max[0],
+              max[1],
+              max[2]),
+          t)) {
+    return std::optional<glm::dvec3>();
+  }
+
+  double tClosest = -1;
+
+  // Support all variations of position component types
+  //
+  // From the glTF spec...
+  // "Floating-point data MUST use IEEE-754 single precision format."
+  //
+  // Yet, the KHR_mesh_quantization extension can specify more
+  //
+  createPositionView(
+      model,
+      positionAccessor,
+      [&transformedRay,
+       &model,
+       &primitive,
+       cullBackFaces,
+       &tClosest,
+       &warnings](const auto& positionView) {
+        // Bail on invalid view
+        if (positionView.status() != AccessorViewStatus::Valid) {
+          warnings.push_back(
+              "Skipping mesh with an invalid position component type");
+          return;
+        }
+
+        bool hasIndexedTriangles = primitive.indices != -1;
+        if (hasIndexedTriangles) {
+          assert(primitive.indices >= 0);
+          const Accessor& indexAccessor =
+              model.accessors[static_cast<size_t>(primitive.indices)];
+
+          // Ignore float index types, these are invalid
+          // From the glTF spec...
+          // "Indices MUST be non-negative integer numbers."
+          if (indexAccessor.componentType == Accessor::ComponentType::FLOAT) {
+            warnings.push_back(
+                "Skipping mesh with an invalid index component type");
+            return;
+          }
+
+          createAccessorView(
+              model,
+              indexAccessor,
+              [&transformedRay,
+               &positionView,
+               &primitive,
+               cullBackFaces,
+               &tClosest,
+               &warnings](const auto& indexView) {
+                // Bail on invalid view
+                if (indexView.status() != AccessorViewStatus::Valid) {
+                  warnings.push_back(
+                      "Could not create accessor view for mesh indices");
+                  return;
+                }
+
+                findClosestIndexedRayHit<
+                    decltype(positionView),
+                    decltype(indexView)>(
```

Comment:
The compiler should be able to deduce these template arguments automatically, no need for this awkward explicit instantiation with decltypes.

Reply:
Good catch. This was leftover from some other in-progress code that eventually got removed.

decltypes are now removed